Upgrade to SD-JWT v12 & API rework

UMR1352 opened the pull request:
Updated the code base to support [SD-JWT version 12](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html) and modified the API to make it simpler and more ergonomic.
## Major changes
- `SdObjectEncoder` / `SdObjectDecoder` are not public anymore, and are only used internally.
- `SdJwt` can only be constructed by parsing a string or through the new `SdJwtBuilder`.
- Added `SdJwtBuilder` which has the same functionalities of `SdObjectEncoder` but:
    - `SdObjectEncoder::conceal` becomes `SdJwtBuilder::make_concealable`.
    - `SdJwtBuilder::finish` returns a whole and valid `SdJwt` instead of a plaintext object and disclosures.
- Added `JwsSigner` trait, to create a JWS from header & payload.
- Added `SdJwtPresentationBuilder` which wraps an `SdJwt` and provides methods to easily conceal disclosable claims and attaching a KeyBinding JWT (KB-JWT).
- Added `KeyBindingJwtBuilder` to simplify the creation of KB-JWTs.
- Added `RequiredKeyBinding` to model a proof of possession for a given key through JWT's `ctf` claim.
## Fixes
- Multibase-base64Url encoding was mistakenly used instead of Base64Url.
- Multiple hashers (to compute disclosure's digests) were allowed but that goes against the spec.

wulfraem commented:
According to [spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#name-object-properties) about the `_sd` array: "The array MAY be empty in case the Issuer decided not to selectively disclose any of the claims at that level. However, it is RECOMMENDED to omit the `_sd` key in this case to save space."

Wanted to test this behavior, but ran into issues, when creating a SD-JWT without any concealable properties. The serialized object had two of the tilde characters directly following another, which should not be the case according to the ABNF (`SD-JWT = JWT "~" *[DISCLOSURE "~"]`). This broke the parsing a step, that comes afterwards in the example. Removing them by hand lead to another error when parsing the SD-JWT without disclosures.

wulfraem reviewed (changes requested):
According to [spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#name-object-properties) about the `_sd` array: "The array MAY be empty in case the Issuer decided not to selectively disclose any of the claims at that level. However, it is RECOMMENDED to omit the `_sd` key in this case to save space."

Wanted to test this behavior, but ran into issues, when creating a SD-JWT without any concealable properties. The serialized object had two of the tilde characters directly following another, which should not be the case according to the ABNF (`SD-JWT = JWT "~" *[DISCLOSURE "~"]`). This broke the parsing a step, that comes afterwards in the example. Removing them by hand lead to another error when parsing the SD-JWT without disclosures.

wulfraem commented:
According to [spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#section-5.2.4.1) about the `_sd` array entries: "The Issuer MUST hide the original order of the claims in the array. To ensure this, it is RECOMMENDED to shuffle the array of hashes, e.g., by sorting it alphanumerically or randomly, after potentially adding decoy digests as described in [Section 5.2.5](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#decoy_digests). The precise method does not matter as long as it does not depend on the original order of elements."

I think, currently the order depends on the order the properties are marked as concealable or decoys are added. The shuffling could be done via calling the builder functions in a random order and only adding one decoy at a time per level, but as the spec also lists sorting the hashes alphabetically, sorting them when adding or serializing might be a nice way to make sure that their order is "shuffled".

wulfraem commented:
Couldn't find a definitive answer in the spec, but when an issuer uses the `.require_key_binding` to add the `cnf` with the ref to the holder, should the holder be able to create a presentation without key binding? Currently, our example does this, but it feels off. If adding the KB-JWT would not be required in this case someone could just strip off the KB-JWT and present the SD-JWT without it, e.g. if not in the possession of the key mentioned in the SD-JWT.

But then again, the [spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#name-key-binding-2) shifts the responsibility for deciding whether the KB-JWT is required or entirely to the verifier, but imho it sounds weird to present a SD-JWT with `cnf` without a KB-JWT.

wulfraem commented:
Could we fix the "erros" -> "errors" typo in the `bug_report.yml` when being at it as well? Don't think, we'll need an extra PR for this. :D

wulfraem commented:
I guess, we'll have to update the CI for the Windows build. Does not seem to be a temporary thing. >.<

UMR1352 commented:
Thank you @wulfraem for being the best reviewer ever. Going through your comments RN.

UMR1352 commented:
> According to [spec](https://www.ietf.org/archive/id/draft-ietf-oauth-selective-disclosure-jwt-12.html#name-object-properties) about the `_sd` array: "The array MAY be empty in case the Issuer decided not to selectively disclose any of the claims at that level. However, it is RECOMMENDED to omit the `_sd` key in this case to save space."
> 
> Wanted to test this behavior, but ran into issues, when creating a SD-JWT without any concealable properties. The serialized object had two of the tilde characters directly following another, which should not be the case according to the ABNF (`SD-JWT = JWT "~" *[DISCLOSURE "~"]`). This broke the parsing a step, that comes afterwards in the example. Removing them by hand lead to another error when parsing the SD-JWT without disclosures.

I fixed this and added a test that makes sure an SD-JWT with 0 disclosure is properly serialized / deserialized even when a KB-JWT is present.

nanderstabel commented on src/sd_jwt.rs:
```diff
@@ -86,15 +189,281 @@ impl FromStr for SdJwt {
+        hasher.alg_name()
+      )));
+    }
+    let disclosures = std::mem::take(&mut sd_jwt.disclosures)
+      .into_iter()
+      .map(|disclosure| (hasher.encoded_digest(disclosure.as_str()), disclosure))
+      .collect();
```
While trying out your other SD-JWT related PR (https://github.com/iotaledger/identity.rs/pull/1413) I ran into the problem that while trying to validate the key binding of a 'freshly' created SD-JWT VC `validate_key_binding` would sometimes randomly fail on [this assertion](https://github.com/iotaledger/identity.rs/blob/a510185d80edfef91a72d675962e23eebb068ab0/identity_credential/src/sd_jwt_vc/token.rs#L372-L374):
```rust
    if kb_jwt.claims().sd_hash != digest {
      return Err(Error::Validation(anyhow!("invalid KB-JWT's `sd_hash`")));
    }
```

I managed to pin the reason down to this piece of code since the disclosures are being collected into a `HashMap` here and therefore being scrambled. That is not necessarily a problem since the spec actually states that it is [RECOMMENDED to shuffle the array of hashes](https://github.com/iotaledger/identity.rs/blob/a510185d80edfef91a72d675962e23eebb068ab0/identity_credential/src/sd_jwt_vc/token.rs#L372-L374) (of the digests). However, the problem is that at this point in the code, the `_sd_hash` has already been generated before the scrambling of the digests which causes `validate_key_binding` to fail at random (because the randomly changed order of the digests causes the `digest` to differ from the `_sd_hash` in `validate_key_binding`).

I could easily fix the problem by changing the `disclosures` type to an `IndexMap` in order to preserve the order of the disclosures:
```rust
#[derive(Debug, Clone)]
pub struct SdJwtPresentationBuilder {
  sd_jwt: SdJwt,
  disclosures: IndexMap<String, Disclosure>,
  removed_disclosures: Vec<Disclosure>,
  object: Value,
}
```

UMR1352 replied:
@nanderstabel thank you for trying this out. This was definitely something I overlooked. I'll update the code with your suggestion
